tiledb/sm/query/readers/result_coords.h

```diff
@@ -240,28 +243,55 @@ struct GlobalOrderResultCoords
         }
       }
 
-      // With bitmap, find the longest contiguous set of bits in the bitmap
-      // from the current position, with coordinares smaller than the next one
-      // in the queue.
-      base::pos_++;
-      while (base::pos_ < cell_num && base::tile_->bitmap_[base::pos_] &&
-             !cmp(*this, next)) {
-        base::pos_++;
-        ret++;
+      // Compute max position.
+      max_pos = base::pos_;
+      while (max_pos < cell_num && base::tile_->bitmap_[max_pos] == 1) {
+        max_pos++;
       }
-    } else {
-      // No bitmap, add all cells from current position, with coordinares
-      // smaller than the next one in the queue.
-      base::pos_++;
-      while (base::pos_ < cell_num - 1 && !cmp(*this, next)) {
-        base::pos_++;
-        ret++;
+      max_pos--;
+    }
+
+    // Find the upper bound comparison by increasing power of two. This ensures
+    // the algorithm works equally well for small slabs vs large ones.
+    uint64_t power_of_two = 1;
+    bool return_max = true;
+    while (return_max && base::pos_ != max_pos) {
```

Comment:
Those 2 new `while` loops are defining 2 algorithms that are quite hard for me to review by reading if they are 100% correct. Are the current UTs covering for all the possible code paths in there?

A way to increase our confidence and readability would be to make a separate function for each of those code blocks so that they can be tested separately and so that their purpose become clearer.

Reply:
This was identified as a bottleneck for SingleCell. It made the merge part of the read in their case go from 2.24879 seconds to 0.0936861, for a ~10 seconds query.

The coverage from the tests is covering all conditions... The new tests is actually running ~10,000 tests, so that should have caught any issues with the search algorithm. Let's discuss offline for a bit.

Reply:
Ok thanks for the explanations. As discussed let's add more documentation to that function to make things clearer.